src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1229 @@
+        assert_eq!(
+            pin_protocol_v1.process_subcommand(&mut rng, &mut persistent_store, client_pin_params),
+            Err(error_code)
+        );
+    }
+
+    #[test]
+    fn test_decrypt_pin() {
+        let shared_secret = [0x88; 32];
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+
+        // "1234"
+        let new_pin_enc = vec![
+            0xC0, 0xCF, 0xAE, 0x4C, 0x79, 0x56, 0x87, 0x99, 0xE5, 0x83, 0x4F, 0xE6, 0x4D, 0xFE,
+            0x53, 0x32, 0x36, 0x0D, 0xF9, 0x1E, 0x47, 0x66, 0x10, 0x5C, 0x63, 0x30, 0x1D, 0xCC,
+            0x00, 0x09, 0x91, 0xA4, 0x20, 0x6B, 0x78, 0x10, 0xFE, 0xC6, 0x2E, 0x7E, 0x75, 0x14,
+            0xEE, 0x01, 0x99, 0x6C, 0xD7, 0xE5, 0x2B, 0xA5, 0x7A, 0x5A, 0xE1, 0xEC, 0x69, 0x31,
+            0x18, 0x35, 0x06, 0x66, 0x97, 0x84, 0x68, 0xC2,
+        ];
+        assert_eq!(
+            decrypt_pin(&aes_dec_key, new_pin_enc),
+            Some(b"1234".to_vec()),
+        );
+
+        // "123"
+        let new_pin_enc = vec![
+            0xF3, 0x54, 0x29, 0x17, 0xD4, 0xF8, 0xCD, 0x23, 0x1D, 0x59, 0xED, 0xE5, 0x33, 0x42,
+            0x13, 0x39, 0x22, 0xBB, 0x91, 0x28, 0x87, 0x6A, 0xF9, 0xB1, 0x80, 0x9C, 0x9D, 0x76,
+            0xFF, 0xDD, 0xB8, 0xD6, 0x8D, 0x66, 0x99, 0xA2, 0x42, 0x67, 0xB0, 0x5C, 0x82, 0x3F,
+            0x08, 0x55, 0x8C, 0x04, 0xC5, 0x91, 0xF0, 0xF9, 0x58, 0x44, 0x00, 0x1B, 0x99, 0xA6,
+            0x7C, 0xC7, 0x2D, 0x43, 0x74, 0x4C, 0x1D, 0x7E,
+        ];
+        assert_eq!(
+            decrypt_pin(&aes_dec_key, new_pin_enc),
+            Some(b"123".to_vec()),
+        );
+
+        // Encrypted PIN is too short.
+        let new_pin_enc = vec![0x44; 63];
+        assert_eq!(decrypt_pin(&aes_dec_key, new_pin_enc), None,);
+
+        // Encrypted PIN is too long.
+        let new_pin_enc = vec![0x44; 65];
+        assert_eq!(decrypt_pin(&aes_dec_key, new_pin_enc), None,);
+    }
+
+    #[test]
+    fn test_check_and_store_new_pin() {
+        let mut rng = ThreadRng256 {};
+        let mut persistent_store = PersistentStore::new(&mut rng);
+        let shared_secret = [0x88; 32];
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+
+        let test_cases = vec![
+            // Accept PIN "1234".
+            (vec![0x31, 0x32, 0x33, 0x34], Ok(())),
+            // Reject PIN "123" since it is too short.
+            (
+                vec![0x31, 0x32, 0x33],
+                Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION),
+            ),
+            // Reject PIN "12'\0'4" (a zero byte at index 2).
+            (
+                vec![0x31, 0x32, 0x00, 0x34],
+                Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION),
+            ),
+            // Reject PINs not ending in 0u8 padding.
```

Comment:
This comment is not 100% clear. The accepted PIN "1234" also doesn't explicitly include 0u8 padding.

Reply:
Changed to a better explanation.

Reply:
I don't see any change uploaded (yet). Please mark as resolved when the code is up-to-date.